File: datasets/C++/Mistral/DataStructures/hard5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>

using namespace std;

vector<vector<int>> computeDistances(int n, const vector<vector<int>>& tree) {
    vector<vector<int>> distance(n + 1, vector<int>(n + 1, 0));

    for (int i = 1; i <= n; ++i) {
        vector<bool> visited(n + 1, false);
        queue<int> q;
        q.push(i);
        visited[i] = true;

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (int v : tree[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    distance[i][v] = distance[i][u] + 1;
                    q.push(v);
                }
            }
        }
    }

    return distance;
}

int main() {
    int n, m;
    cout << "Enter the number of vertices (n): ";
    cin >> n;
    cout << "Enter the number of points (m): ";
    cin >> m;

    vector<vector<int>> tree(n + 1);
    cout << "Enter the edges (u and v for each edge):" << endl;
    for (int i = 0; i < n - 1; ++i) {
        int u, v;
        cout << "Enter edge " << i + 1 << " (u v): ";
        cin >> u >> v;
        tree[u].push_back(v);
        tree[v].push_back(u);
    }

    vector<vector<int>> distance = computeDistances(n, tree);

    vector<pair<int, int>> points(m);
    cout << "Enter the points (x and y for each point):" << endl;
    for (int i = 0; i < m; ++i) {
        cout << "Enter point " << i + 1 << " (x y): ";
        cin >> points[i].first >> points[i].second;
    }

    int max_distance = 0;
    for (int i = 0; i < m; ++i) {
        for (int j = i + 1; j < m; ++j) {
            int current_distance = distance[points[i].first][points[j].first] +
                                  distance[points[i].second][points[j].second];
            if (current_distance > max_distance) {
                max_distance = current_distance;
            }
        }
    }

    cout << "The maximum distance is: " << max_distance << endl;

    return 0;
}

File: datasets/Python/Mistral/DataStructures/hard5.py
from collections import deque

def main():
    # Read n and m interactively
    n = int(input("Enter the number of vertices (n): "))
    m = int(input("Enter the number of points (m): "))

    # Build the tree
    tree = [[] for _ in range(n + 1)]
    print("Enter the edges (u and v for each edge):")
    for i in range(n - 1):
        u, v = map(int, input(f"Enter edge {i + 1} (u v): ").split())
        tree[u].append(v)
        tree[v].append(u)

    # Precompute distances between all pairs of vertices using BFS
    distance = [[0] * (n + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        queue = deque([i])
        visited = [False] * (n + 1)
        visited[i] = True
        while queue:
            u = queue.popleft()
            for v in tree[u]:
                if not visited[v]:
                    visited[v] = True
                    distance[i][v] = distance[i][u] + 1
                    queue.append(v)

    # Process each point and find the maximum distance
    max_distance = 0
    points = []
    print("Enter the points (x and y for each point):")
    for i in range(m):
        x, y = map(int, input(f"Enter point {i + 1} (x y): ").split())
        points.append((x, y))

    for i in range(m):
        xi, yi = points[i]
        for j in range(i + 1, m):
            xj, yj = points[j]
            current_distance = distance[xi][xj] + distance[yi][yj]
            if current_distance > max_distance:
                max_distance = current_distance

    print("The maximum distance is:", max_distance)

if __name__ == "__main__":
    main()